### PR TITLE
Day/Night Time Cycle For Maps

### PR DESCRIPTION
**Pull request by DocNITE**
# Description

Add real world Day/Night cycle changing, with Ambient color interpolation and some configurable features for mappers. 
High inspired by [Nuclear-14](https://github.com/Vault-Overseers/nuclear-14.git) and [Crystall Punk 14](https://github.com/crystallpunk-14/)

TimeCycle use 24h day calculation, instead coefficient, like in N14. 

---

# TODO

[ ] - Add more examples/prepared map palettes for Day/Night ambient.
[ ] - (Optional) Move all palette description into List of variables, instead calling `IPrototypeManager` every minute passing.

---

https://github.com/user-attachments/assets/cc5d1b8c-bd25-4042-8cee-e2edfa0d9acc

image

---

# Changelog

:cl:
- add: Added Day/Night time cycle for admins and mapers.

### CODE REVIEW

**Comment by DocNITE** [comment]
I don't sure - should i calculate day cycle in coefficient/percentage. So, if you have any idea - "tell" me :eyes:

File: Content.Shared/TimeCycle/TimeCycleComponent.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+    [DataField, ViewVariables(VVAccess.ReadWrite)]
+    public TimeSpan SpeedUpMinuteDuration { get; set; } = TimeSpan.FromMilliseconds(10);
+
+    // NOTE: Default time should be is noon
+    [DataField, ViewVariables(VVAccess.ReadWrite)]
+    public TimeSpan CurrentTime { get; set; } = TimeSpan.FromHours(12);
+
+    [DataField, ViewVariables(VVAccess.ReadWrite)]
+    public string PalettePrototype = "DefaultTimeCycle";
```
Comment (VMSolidus) [inline Content.Shared/TimeCycle/TimeCycleComponent.cs:26]:
```suggestion
    [DataField]
    public bool SpeedUp;

    [DataField]
    public bool Paused;

    [DataField]
    public TimeSpan MinuteDuration { get; set; } = TimeSpan.FromSeconds(4);

    [DataField]
    public TimeSpan SpeedUpMinuteDuration { get; set; } = TimeSpan.FromMilliseconds(10);

    // NOTE: Default time should be is noon
    [DataField]
    public TimeSpan CurrentTime { get; set; } = TimeSpan.FromHours(12);

    [DataField]
    public string PalettePrototype = "DefaultTimeCycle";
```

File: Content.Server/TimeCycle/TimeCycleSystem.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    {
+        var curTime = _gameTiming.CurTime;
+        var query = EntityQueryEnumerator<TimeCycleComponent, MapLightComponent>();
+
+        while (query.MoveNext(out var mapid, out var timeComp, out var mapLightComp))
+        {
+            if (timeComp.Paused)
+                continue;
+            if (curTime < timeComp.DelayTime)
+                continue;
```
Comment (VMSolidus) [inline Content.Server/TimeCycle/TimeCycleSystem.cs:28]:
```suggestion
            if (timeComp.Paused
                || curTime < timeComp.DelayTime)
                continue;
```

File: Content.Server/TimeCycle/TimeCycleSystem.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+            // Should be used for developing time palletes or for debuging
+            // O-o-or... You can cosplay pucchi from JoJo 6 with his 'Made In Heaven'
+            if (timeComp.SpeedUp)
+                timeComp.DelayTime = curTime + timeComp.SpeedUpMinuteDuration;
+            else
+                timeComp.DelayTime = curTime + timeComp.MinuteDuration;
```
Comment (VMSolidus) [inline Content.Server/TimeCycle/TimeCycleSystem.cs:35]:
```suggestion
            timeComp.DelayTime = curTime + (timeComp.SpeedUp ? timeComp.SpeedUpMinuteDuration : timeComp.MinuteDuration);
```

File: Content.Server/TimeCycle/TimeCycleSystem.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+        base.Update(frameTime);
+    }
+
+    private void UpdateAmbientColor(EntityUid mapid, TimeCycleComponent timeComp, MapLightComponent mapLightComp)
+    {
+        if (!_prototypeManager.TryIndex(timeComp.PalettePrototype, out TimeCyclePalettePrototype? timeEntries))
+            return;
+        if (timeEntries is null)
+            return;
```
Comment (VMSolidus) [inline Content.Server/TimeCycle/TimeCycleSystem.cs:52]:
```suggestion
        if (!_prototypeManager.TryIndex(timeComp.PalettePrototype, out TimeCyclePalettePrototype? timeEntries)
            || timeEntries is null)
            return;
```

File: Content.Server/TimeCycle/TimeCycleSystem.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+    // We should convert current 'TimeSpan' (with days) time into one day cycle time (in 24 hours)
+    private TimeSpan GetTimeInCycle(TimeSpan timeSpan)
+    {
+        double timeCycleMilliseconds = TimeSpan.FromHours(24).TotalMilliseconds;
+        double totalMilliseconds = timeSpan.TotalMilliseconds;
+        double timeInCycleMilliseconds = totalMilliseconds % timeCycleMilliseconds;
+
+        return TimeSpan.FromMilliseconds(timeInCycleMilliseconds);;
+    }
```
Comment (VMSolidus) [inline Content.Server/TimeCycle/TimeCycleSystem.cs:67]:
```suggestion
    private TimeSpan GetTimeInCycle(TimeSpan timeSpan) =>
        timeSpan.TotalMilliseconds % TimeSpan.FromHours(24).TotalMilliseconds;
```

**Review by VMSolidus — CHANGES_REQUESTED** [review]
Yep, pretty simple changes to do here.

**Review by VMSolidus — APPROVED** [review]

**Pull request merged**